components/concurrency_manager/src/lib.rs

```diff
@@ -52,9 +145,119 @@ impl ConcurrencyManager {
+                // limit is valid
+                let source = source.into_error_source();
+                self.report_error(new_ts, limit.limit, source, true)?;
+            } else {
+                // limit is stale
+                // use an approximate limit to avoid false alerts caused by failed limit updates
+
+                let approximate_limit = TimeStamp::compose(
+                    limit.limit.physical() + duration_to_last_limit_update.as_millis() as u64,
+                    limit.limit.logical(),
+                );
+
+                if new_ts > approximate_limit {
+                    let source = source.into_error_source();
+                    self.report_error(new_ts, approximate_limit, source, false)?;
+                }
+            }
+        }
+
+        MAX_TS_GAUGE.set(
+            self.max_ts
+                .fetch_max(new_ts.into_inner(), Ordering::SeqCst)
+                .max(new_ts.into_inner()) as i64,
+        );
+        Ok(())
+    }
+
+    fn report_error(
+        &self,
+        new_ts: TimeStamp,
+        limit: TimeStamp,
+        source: impl slog::Value + Display,
+        can_panic: bool,
+    ) -> Result<(), InvalidMaxTsUpdate> {
+        error!("invalid max_ts update";
+            "attempted_ts" => new_ts,
+            "max_allowed" => limit.into_inner(),
+            "source" => &source,
+        );
+        match self.action_on_invalid_max_ts.load() {
+            ActionOnInvalidMaxTs::Panic if can_panic => {
+                panic!(
+                    "invalid max_ts update: {} exceeds the limit {}, source={}",
+                    new_ts,
+                    limit.into_inner(),
+                    source
+                );
+            }
+            ActionOnInvalidMaxTs::Error => Err(InvalidMaxTsUpdate {
+                attempted_ts: new_ts,
+                max_allowed: limit,
+            }),
+            ActionOnInvalidMaxTs::Log => Ok(()),
+            ActionOnInvalidMaxTs::Panic => Ok(()),
+        }
+    }
+
+    /// Set the maximum allowed value for max_ts updates, except for the updates
+    /// from PD TSO. The limit must be updated regularly to prevent the
+    /// blocking of max_ts. It prevents max_ts from being updated to an
+    /// unreasonable value, which is usually caused by bugs or unsafe
+    /// usages.
+    ///
+    /// # Note
+    /// If the new limit is smaller than the current limit, this operation will
+    /// have no effect and return silently.
+    pub fn set_max_ts_limit(&self, limit: TimeStamp) {
+        if limit.is_max() {
+            error!("max_ts_limit cannot be set to u64::max");
+            return;
+        }
+
+        let current = self.max_ts_limit.load();
+        if limit.into_inner() > current.limit.into_inner() {
+            let new_state = MaxTsLimit {
+                limit,
+                update_time: self.time_provider.now(),
+            };
+            self.max_ts_limit.store(new_state);
```

Comment:
Should we consider using a CAS loop here?

Reply:
A race is not likely to happen here, but it won't hurt.